llama_index/chat_engine/types.py

```diff
@@ -1,14 +1,91 @@
 
-from llama_index.llms.base import ChatMessage
-from llama_index.response.schema import RESPONSE_TYPE
+from llama_index.llms.base import ChatMessage, ChatResponseGen, ChatResponseAsyncGen
+from llama_index.response.schema import RESPONSE_TYPE, StreamingResponse
 
 logger = logging.getLogger(__name__)
 
 
+class StreamingChatResponse:
+    """Streaming chat response to user and writing to chat history."""
+
+    def __init__(
+        self, chat_stream: Union[ChatResponseGen, ChatResponseAsyncGen]
+    ) -> None:
+        self._chat_stream = chat_stream
+        self._queue: queue.Queue = queue.Queue()
+        self._is_done = False
+        self._is_function: Optional[bool] = None
+        self.response_str = ""
+
+    def __str__(self) -> str:
+        if self._is_done and not self._queue.empty() and not self._is_function:
+            for delta in self._queue.queue:
+                self.response_str += delta
+        return self.response_str
+
+    def write_response_to_history(self, chat_history: List[ChatMessage]) -> None:
+        if isinstance(self._chat_stream, AsyncGenerator):
+            raise ValueError(
+                "Cannot write to history with async generator in sync function."
+            )
+
+        final_message = None
+        for chat in self._chat_stream:
+            final_message = chat.message
+            self._is_function = (
+                final_message.additional_kwargs.get("function_call", None) is not None
+            )
+            self._queue.put_nowait(chat.delta)
+
+        if final_message is not None:
+            chat_history.append(final_message)
+
+        self._is_done = True
+
+    async def awrite_response_to_history(self, chat_history: List[ChatMessage]) -> None:
+        if isinstance(self._chat_stream, Generator):
+            raise ValueError(
+                "Cannot write to history with sync generator in async function."
+            )
+
+        final_message = None
+        async for chat in self._chat_stream:
+            final_message = chat.message
+            self._is_function = (
+                final_message.additional_kwargs.get("function_call", None) is not None
+            )
+            self._queue.put_nowait(chat.delta)
+
+        if final_message is not None:
+            chat_history.append(final_message)
+
+        self._is_done = True
+
+    @property
+    def response_gen(self) -> Generator[str, None, None]:
+        while not self._is_done or not self._queue.empty():
+            try:
+                delta = self._queue.get(block=False)
+                self.response_str += delta
+                yield delta
+            except queue.Empty:
+                # Queue is empty, but we're not done yet
+                continue
+
+
+STREAMING_CHAT_RESPONSE_TYPE = Union[
+    StreamingResponse,
+    StreamingChatResponse,
+    Generator[StreamingChatResponse, None, None],
+    AsyncGenerator[StreamingChatResponse, None],
+]
```

Comment:
imo a little strange that we lump all of these into a union type.
especially for the Streaming* objects, you need to call the response_gen to get the generators, but the other types are already generator or async generators.

Reply:
will have to rethink this in future PRs.

Reply:
I think we need to add docs on why an open ai agent returns a generator of generators, that's not super obvious to a user.

Reply:
Yea, it's just because there can be multiple steps. I do showcase it in a notebook, but will also find a place in the docs to state this.

Reply:
Part of the reason here is that the agent and chat engine both share the same base class, but have pretty drastically different ways of streaming